# Problem Statement:
  Fix Dhrystone degradation with GCC v492.
  Target: e5500 64-bit.

# Owned by:
  Rohit

# Action:
  * Adjust the address cost while accessing constant addresses
    so that it is profitable for loop invariant code.

  * Referenced test case part of dhrystone benchmark code.

diff -Naur gcc-4.9.2/gcc/config/rs6000/rs6000.c gcc-4.9.2-patch/gcc/config/rs6000/rs6000.c
--- gcc-4.9.2/gcc/config/rs6000/rs6000.c	2014-10-12 21:33:20.000000000 -0500
+++ gcc-4.9.2-patch/gcc/config/rs6000/rs6000.c	2014-11-22 13:38:52.939563569 -0600
@@ -1460,7 +1460,7 @@
 #undef TARGET_RTX_COSTS
 #define TARGET_RTX_COSTS rs6000_rtx_costs
 #undef TARGET_ADDRESS_COST
-#define TARGET_ADDRESS_COST hook_int_rtx_mode_as_bool_0
+#define TARGET_ADDRESS_COST rs6000_address_costs
 
 #undef TARGET_DWARF_REGISTER_SPAN
 #define TARGET_DWARF_REGISTER_SPAN rs6000_dwarf_register_span
@@ -29899,6 +29899,24 @@
   return ret;
 }
 
+/* Address cost calculation.
+ * FIXME: Handle all addressing types */
+
+static int
+rs6000_address_costs (rtx x, enum machine_mode mode ATTRIBUTE_UNUSED,
+		   addr_space_t as ATTRIBUTE_UNUSED,
+		   bool speed ATTRIBUTE_UNUSED)
+{
+  if (toc_relative_expr_p (x, false))
+    {
+      rtx temp = XVECEXP (tocrel_base, 0, 0);
+      if (GET_CODE (temp) != SYMBOL_REF)
+        return 0;
+      return 5;
+    }
+  return 0; 
+}
+
 /* Debug form of ADDRESS_COST that is selected if -mdebug=cost.  */
 
 static int
diff -Naur gcc-4.9.2/gcc/testsuite/gcc.target/powerpc/ppc-fsl-loop-inv-2.c gcc-4.9.2-patch/gcc/testsuite/gcc.target/powerpc/ppc-fsl-loop-inv-2.c
--- gcc-4.9.2/gcc/testsuite/gcc.target/powerpc/ppc-fsl-loop-inv-2.c	1969-12-31 18:00:00.000000000 -0600
+++ gcc-4.9.2-patch/gcc/testsuite/gcc.target/powerpc/ppc-fsl-loop-inv-2.c	2015-01-22 16:07:13.614626997 -0600
@@ -0,0 +1,292 @@
+/* { dg-do compile { target { powerpc*-*-* && lp64 } } } */
+/* { dg-options "-O3 -m64 -fdump-rtl-loop2_invariant" } */
+/* { dg-final { scan-rtl-dump-times "Decided to move" 16 "loop2_invariant"} } */
+/* { dg-final { cleanup-rtl-dump "loop2_invariant" } } */
+
+#include <stdio.h>
+#include <string.h>
+
+typedef enum {Ident_1, Ident_2, Ident_3, Ident_4, Ident_5}
+                Enumeration;
+typedef int One_Thirty;
+typedef int One_Fifty;
+typedef char Capital_Letter;
+typedef int Boolean;
+typedef char Str_30 [31];
+typedef int Arr_1_Dim [50];
+typedef int Arr_2_Dim [50] [50];
+
+typedef struct record
+{
+  struct record *Ptr_Comp;
+  Enumeration Discr;
+  union {
+    struct {
+      Enumeration Enum_Comp;
+      int Int_Comp;
+      char Str_Comp [31];
+    } var_1;
+    struct {
+      Enumeration E_Comp_2;
+      char Str_2_Comp [31];
+    } var_2;
+    struct {
+      char Ch_1_Comp;
+      char Ch_2_Comp;
+    } var_3;
+  } variant;
+} Rec_Type, *Rec_Pointer;
+
+Rec_Pointer Ptr_Glob, Next_Ptr_Glob;
+int Int_Glob;
+Boolean Bool_Glob;
+char Ch_1_Glob, Ch_2_Glob;
+int Arr_1_Glob [50];
+int Arr_2_Glob [50] [50];
+char Reg_Define[] = "Register option selected.";
+double Begin_Time, End_Time, User_Time;
+double Microseconds, Dhrystones_Per_Second, Vax_Mips;
+
+Enumeration Func_1 ();
+void exit (int);
+
+void main ()
+{
+  double dtime();
+  One_Fifty Int_1_Loc;
+  One_Fifty Int_2_Loc;
+  One_Fifty Int_3_Loc;
+  char Ch_Index;
+  Enumeration Enum_Loc;
+  Str_30 Str_1_Loc;
+  Str_30 Str_2_Loc;
+  int Run_Index;
+  int Number_Of_Runs;
+
+  FILE *Ap;
+
+
+
+  if ((Ap = fopen("dhry.res","a+")) == ((void *)0))
+    {
+       printf("Can not open dhry.res\n\n");
+       exit(1);
+    }
+
+  Next_Ptr_Glob = (Rec_Pointer) malloc (sizeof (Rec_Type));
+  Ptr_Glob = (Rec_Pointer) malloc (sizeof (Rec_Type));
+
+  Ptr_Glob->Ptr_Comp = Next_Ptr_Glob;
+  Ptr_Glob->Discr = Ident_1;
+  Ptr_Glob->variant.var_1.Enum_Comp = Ident_3;
+  Ptr_Glob->variant.var_1.Int_Comp = 40;
+  strcpy (Ptr_Glob->variant.var_1.Str_Comp,
+          "DHRYSTONE PROGRAM, SOME STRING");
+  strcpy (Str_1_Loc, "DHRYSTONE PROGRAM, 1'ST STRING");
+
+  Arr_2_Glob [8][7] = 10;
+
+  printf ("\n");
+  printf ("Dhrystone Benchmark, Version 2.1 (Language: C)\n");
+  printf ("\n");
+  printf ("Please give the number of runs through the benchmark: ");
+  {
+    int n;
+    scanf ("%d", &n);
+    Number_Of_Runs = n;
+  }
+  printf ("\n");
+  printf ("Execution starts, %d runs through Dhrystone\n",Number_Of_Runs);
+
+  Begin_Time = dtime();
+  for (Run_Index = 1; Run_Index <= Number_Of_Runs; ++Run_Index)
+  {
+
+    Proc_5();
+    Proc_4();
+
+    Int_1_Loc = 2;
+    Int_2_Loc = 3;
+    strcpy (Str_2_Loc, "DHRYSTONE PROGRAM, 2'ND STRING");
+    Enum_Loc = Ident_2;
+    Bool_Glob = ! Func_2 (Str_1_Loc, Str_2_Loc);
+
+    while (Int_1_Loc < Int_2_Loc)
+    {
+      Int_3_Loc = 5 * Int_1_Loc - Int_2_Loc;
+      Proc_7 (Int_1_Loc, Int_2_Loc, &Int_3_Loc);
+      Int_1_Loc += 1;
+    }
+
+    Proc_8 (Arr_1_Glob, Arr_2_Glob, Int_1_Loc, Int_3_Loc);
+
+    Proc_1 (Ptr_Glob);
+    for (Ch_Index = 'A'; Ch_Index <= Ch_2_Glob; ++Ch_Index)
+    {
+      if (Enum_Loc == Func_1 (Ch_Index, 'C'))
+
+        {
+          Proc_6 (Ident_1, &Enum_Loc);
+          strcpy (Str_2_Loc, "DHRYSTONE PROGRAM, 3'RD STRING");
+          Int_2_Loc = Run_Index;
+          Int_Glob = Run_Index;
+        }
+    }
+    Int_2_Loc = Int_2_Loc * Int_1_Loc;
+    Int_1_Loc = Int_2_Loc / Int_3_Loc;
+    Int_2_Loc = 7 * (Int_2_Loc - Int_3_Loc) - Int_1_Loc;
+    Proc_2 (&Int_1_Loc);
+  }
+  End_Time = dtime();
+
+  printf ("Execution ends\n");
+  printf ("\n");
+  printf ("Final values of the variables used in the benchmark:\n");
+  printf ("\n");
+  printf ("Int_Glob:            %d\n", Int_Glob);
+  printf ("        should be:   %d\n", 5);
+  printf ("Bool_Glob:           %d\n", Bool_Glob);
+  printf ("        should be:   %d\n", 1);
+  printf ("Ch_1_Glob:           %c\n", Ch_1_Glob);
+  printf ("        should be:   %c\n", 'A');
+  printf ("Ch_2_Glob:           %c\n", Ch_2_Glob);
+  printf ("        should be:   %c\n", 'B');
+  printf ("Arr_1_Glob[8]:       %d\n", Arr_1_Glob[8]);
+  printf ("        should be:   %d\n", 7);
+  printf ("Arr_2_Glob[8][7]:    %d\n", Arr_2_Glob[8][7]);
+  printf ("        should be:   Number_Of_Runs + 10\n");
+  printf ("Ptr_Glob->\n");
+  printf ("  Ptr_Comp:          %d\n", Ptr_Glob->Ptr_Comp);
+  printf ("        should be:   (implementation-dependent)\n");
+  printf ("  Discr:             %d\n", Ptr_Glob->Discr);
+  printf ("        should be:   %d\n", 0);
+  printf ("  Enum_Comp:         %d\n", Ptr_Glob->variant.var_1.Enum_Comp);
+  printf ("        should be:   %d\n", 2);
+  printf ("  Int_Comp:          %d\n", Ptr_Glob->variant.var_1.Int_Comp);
+  printf ("        should be:   %d\n", 17);
+  printf ("  Str_Comp:          %s\n", Ptr_Glob->variant.var_1.Str_Comp);
+  printf ("        should be:   DHRYSTONE PROGRAM, SOME STRING\n");
+  printf ("Next_Ptr_Glob->\n");
+  printf ("  Ptr_Comp:          %d\n", Next_Ptr_Glob->Ptr_Comp);
+  printf ("        should be:   (implementation-dependent), same as above\n");
+  printf ("  Discr:             %d\n", Next_Ptr_Glob->Discr);
+  printf ("        should be:   %d\n", 0);
+  printf ("  Enum_Comp:         %d\n", Next_Ptr_Glob->variant.var_1.Enum_Comp);
+  printf ("        should be:   %d\n", 1);
+  printf ("  Int_Comp:          %d\n", Next_Ptr_Glob->variant.var_1.Int_Comp);
+  printf ("        should be:   %d\n", 18);
+  printf ("  Str_Comp:          %s\n", Next_Ptr_Glob->variant.var_1.Str_Comp);
+  printf ("        should be:   DHRYSTONE PROGRAM, SOME STRING\n");
+  printf ("Int_1_Loc:           %d\n", Int_1_Loc);
+  printf ("        should be:   %d\n", 5);
+  printf ("Int_2_Loc:           %d\n", Int_2_Loc);
+  printf ("        should be:   %d\n", 13);
+  printf ("Int_3_Loc:           %d\n", Int_3_Loc);
+  printf ("        should be:   %d\n", 7);
+  printf ("Enum_Loc:            %d\n", Enum_Loc);
+  printf ("        should be:   %d\n", 1);
+  printf ("Str_1_Loc:           %s\n", Str_1_Loc);
+  printf ("        should be:   DHRYSTONE PROGRAM, 1'ST STRING\n");
+  printf ("Str_2_Loc:           %s\n", Str_2_Loc);
+  printf ("        should be:   DHRYSTONE PROGRAM, 2'ND STRING\n");
+  printf ("\n");
+
+  User_Time = End_Time - Begin_Time;
+
+  if (User_Time < 2)
+  {
+    printf ("Measured time too small to obtain meaningful results\n");
+    printf ("Please increase number of runs\n");
+    printf ("\n");
+  }
+  else
+  {
+    Microseconds = User_Time * 1000000.0
+                        / (double) Number_Of_Runs;
+    Dhrystones_Per_Second = (double) Number_Of_Runs / User_Time;
+    Vax_Mips = Dhrystones_Per_Second / 1757.0;
+
+    printf ("Register option selected?  NO\n");
+    strcpy(Reg_Define, "Register option not selected.");
+
+    printf ("Microseconds for one run through Dhrystone: ");
+    printf ("%7.1lf \n", Microseconds);
+    printf ("Dhrystones per Second:                      ");
+    printf ("%10.1lf \n", Dhrystones_Per_Second);
+    printf ("VAX MIPS rating = %10.3lf \n",Vax_Mips);
+    printf ("\n");
+
+    fprintf(Ap,"\n");
+    fprintf(Ap,"Dhrystone Benchmark, Version 2.1 (Language: C)\n");
+    fprintf(Ap,"%s\n",Reg_Define);
+    fprintf(Ap,"Microseconds for one loop: %7.1lf\n",Microseconds);
+    fprintf(Ap,"Dhrystones per second: %10.1lf\n",Dhrystones_Per_Second);
+    fprintf(Ap,"VAX MIPS rating: %10.3lf\n",Vax_Mips);
+    fclose(Ap);
+  }
+}
+
+Proc_1 (Ptr_Val_Par)
+Rec_Pointer Ptr_Val_Par;
+{
+  Rec_Pointer Next_Record = Ptr_Val_Par->Ptr_Comp;
+  *Ptr_Val_Par->Ptr_Comp = *Ptr_Glob;
+  Ptr_Val_Par->variant.var_1.Int_Comp = 5;
+  Next_Record->variant.var_1.Int_Comp
+        = Ptr_Val_Par->variant.var_1.Int_Comp;
+  Next_Record->Ptr_Comp = Ptr_Val_Par->Ptr_Comp;
+  Proc_3 (&Next_Record->Ptr_Comp);
+
+  if (Next_Record->Discr == Ident_1)
+  {
+    Next_Record->variant.var_1.Int_Comp = 6;
+    Proc_6 (Ptr_Val_Par->variant.var_1.Enum_Comp,
+           &Next_Record->variant.var_1.Enum_Comp);
+    Next_Record->Ptr_Comp = Ptr_Glob->Ptr_Comp;
+    Proc_7 (Next_Record->variant.var_1.Int_Comp, 10,
+           &Next_Record->variant.var_1.Int_Comp);
+  }
+  else
+    *Ptr_Val_Par = *Ptr_Val_Par->Ptr_Comp;
+}
+
+
+Proc_2 (Int_Par_Ref)
+One_Fifty *Int_Par_Ref;
+{
+  One_Fifty Int_Loc;
+  Enumeration Enum_Loc;
+  Int_Loc = *Int_Par_Ref + 10;
+  do
+    if (Ch_1_Glob == 'A')
+    {
+      Int_Loc -= 1;
+      *Int_Par_Ref = Int_Loc - Int_Glob;
+      Enum_Loc = Ident_1;
+    }
+  while (Enum_Loc != Ident_1);
+}
+
+
+Proc_3 (Ptr_Ref_Par)
+Rec_Pointer *Ptr_Ref_Par;
+
+{
+  if (Ptr_Glob != 0)
+    *Ptr_Ref_Par = Ptr_Glob->Ptr_Comp;
+  Proc_7 (10, Int_Glob, &Ptr_Glob->variant.var_1.Int_Comp);
+}
+
+Proc_4 ()
+{
+  Boolean Bool_Loc;
+  Bool_Loc = Ch_1_Glob == 'A';
+  Bool_Glob = Bool_Loc | Bool_Glob;
+  Ch_2_Glob = 'B';
+}
+
+Proc_5 ()
+{
+  Ch_1_Glob = 'A';
+  Bool_Glob = 0;
+}
